source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,408 @@
+// a matching LbSubsetEnryPtr, if any.
+SubsetLoadBalancer::LbSubsetEntryPtr SubsetLoadBalancer::findSubset(
+    const std::vector<Router::MetadataMatchCriterionConstSharedPtr>& match_criteria) {
+  const LbSubsetMap* subsets = &subsets_;
+
+  // Because the match_criteria and the host metadata used to populate subsets_ are sorted in the
+  // same order, we can iterate over the criteria and perform a lookup for each key and value,
+  // starting with the root LbSubsetMap and using the previous iteration's LbSubsetMap thereafter
+  // (tracked in subsets). If ever a criterion's key or value is not found, there is no subset for
+  // this criteria. If we reach the last criterion, we've found the LbSubsetEntry for the criteria,
+  // which may or may not have a subset attached to it.
+  for (uint32_t i = 0; i < match_criteria.size(); i++) {
+    const Router::MetadataMatchCriterion& match_criterion = *match_criteria[i];
+    const auto& subset_it = subsets->find(match_criterion.name());
+    if (subset_it == subsets->end()) {
+      // No subsets with this key (at this level in the hierarchy).
+      break;
+    }
+
+    const ValueSubsetMap& vs_map = subset_it->second;
+    const auto& vs_it = vs_map.find(match_criterion.value());
+    if (vs_it == vs_map.end()) {
+      // No subsets with this value.
+      break;
+    }
+
+    const LbSubsetEntryPtr& entry = vs_it->second;
+    if (i + 1 == match_criteria.size()) {
+      // We've reached the end of the criteria, and they all matched.
+      return entry;
+    }
+
+    subsets = &entry->children_;
+  }
+
+  return nullptr;
+}
+
+void SubsetLoadBalancer::updateFallbackSubset(const std::vector<HostSharedPtr>& hosts_added,
+                                              const std::vector<HostSharedPtr>& hosts_removed) {
+  if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::NO_FALLBACK) {
+    return;
+  }
+
+  HostPredicate predicate;
+
+  if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::ANY_ENDPOINT) {
+    predicate = [](const Host&) -> bool { return true; };
+  } else {
+    predicate =
+        std::bind(&SubsetLoadBalancer::hostMatchesDefaultSubset, this, std::placeholders::_1);
+  }
+
+  if (fallback_subset_ == nullptr) {
+    // First update: create the default host subset.
+    fallback_subset_.reset(new LbSubsetEntry());
+    fallback_subset_->initLoadBalancer(*this, predicate);
+  } else {
+    // Subsequent updates: add/remove hosts.
+    fallback_subset_->host_subset_->update(hosts_added, hosts_removed, predicate);
+  }
+}
+
+// Given the addition and/or removal of hosts, update all subsets, creating new subsets as
+// necessary.
+void SubsetLoadBalancer::update(const std::vector<HostSharedPtr>& hosts_added,
+                                const std::vector<HostSharedPtr>& hosts_removed) {
+  updateFallbackSubset(hosts_added, hosts_removed);
+
+  std::unordered_set<LbSubsetEntryPtr> subsets_modified;
```

Comment:
random thought: 151-157 is the same in the added/removed case. Could have some type of function which takes a lambda which then gets called back with with host/subset pairs to act on. Might make the actual logic in this function a bit easier to read. Take it or leave it.